### PR TITLE
Implement sponge execution inside Keccak interpreter

### PR DESCRIPTION
**Pull request by querolita**
Includes `run_sponge`, `run_squeeze` and `run_absorb`.

### CODE REVIEW

File: optimism/src/keccak/witness.rs [hunk 1]
```diff
@@ -83,15 +83,76 @@ impl<Fp: Field> KeccakInterpreter for KeccakEnv<Fp> {
...
+        // Rest is zero thanks to null_state
+
+        // TODO: more updates to the env?
+    }
+
+    fn run_absorb(&mut self, absorb: Absorb) {
+        self.set_flag_absorb(absorb);
+
+        // Compute witness values
+        let ini_idx = self.block_idx * RATE_IN_BYTES;
```
Comment (dannywillems) [inline optimism/src/keccak/witness.rs:122]: The drawing I did during our call. Only if you want to add it.
```
        //    Permutation + Mode of Operation (Sponge)
        //    state -> permutation(state) -> state'
        //              ----> either [0] or state'
        //             |            new state = Exp(block)
        //             |         ------------------------
        //    Absorb: state  + [  block      +     0...0 ]
        //                       1088 bits          512
        //            ----------------------------------
        //                         XOR STATE
```

Comment (dannywillems) [reply optimism/src/keccak/witness.rs:122]: Nit (not blocking, and only a personal taste): I would simply reorganize the variable definition in a different order
```
        let ini_idx = self.block_idx * RATE_IN_BYTES;
        let mut block = self.padded[ini_idx..ini_idx + RATE_IN_BYTES].to_vec();
        // Pad with zeros
        block.append(&mut vec![0; CAPACITY_IN_BYTES]);

        let old_state = self.prev_block.clone();
        let new_state = Keccak::expand_state(&block);
        let xor_state = old_state
            .iter()
            .zip(new_state.clone())
            .map(|(x, y)| x + y)
            .collect::<Vec<u64>>();
```

File: optimism/src/keccak/witness.rs [hunk 2]
```diff
@@ -83,15 +83,76 @@ impl<Fp: Field> KeccakInterpreter for KeccakEnv<Fp> {
...
+        let xor_state = old_state
+            .iter()
+            .zip(new_state.clone())
+            .map(|(x, y)| x + y)
+            .collect::<Vec<u64>>();
+
+        // Write absorb-related columns
+        for i in 0..QUARTERS * DIM * DIM {
+            self.write_column(KeccakColumn::SpongeOldState(i), old_state[i]);
+            self.write_column(KeccakColumn::SpongeNewState(i), new_state[i]);
```
Comment (dannywillems) [inline optimism/src/keccak/witness.rs:139]: As discussed, maybe it can be simplified later as the new state is only used temporally. Opening a follow-up.

**Review by dannywillems — APPROVED** [review]

File: optimism/src/keccak/witness.rs (hunk 1 above)
Comment (querolita) [reply optimism/src/keccak/witness.rs:122]: I included them in https://github.com/o1-labs/proof-systems/pull/1559/

**Pull request merged**